### PR TITLE
Consolidate READ/INSERT/DELETE REPORT statements

### PR DESCRIPTION
**Pull request by mbtools**
Preparation for #6154, no functional changes

Add class `zcl_abapgit_sap_report` to handle all situations dealing directly with report sources. This provides a central place to handle ABAP language version.

Includes factory/injector to facilitate testing and open-abap.

Ref #6298

Tests:

image
image
image

### CODE REVIEW

**Comment by mbtools** [comment]
Is there a way to exclude `zcl_abapgit_sap_report` from the tests, @larshp?

**Comment by larshp** [comment]
will add scaffolding for the statement in https://github.com/abaplint/transpiler/pull/1250, so it compiles the statement, but fails if the unit tests hits it

File: src/objects/sap/zif_abapgit_sap_report.intf.abap [hunk 1]
```diff
@@ -0,0 +1,54 @@
+INTERFACE zif_abapgit_sap_report
+  PUBLIC.
+
+  TYPES:
+    ty_abap_language_version TYPE c LENGTH 1.
+
+  METHODS read_report
+    IMPORTING
+      iv_name          TYPE programm
```
Comment (larshp) [inline src/objects/sap/zif_abapgit_sap_report.intf.abap:9]: for the `PROGRAMM`, can https://abapedia.org/steampunk-2305-api/syrepid.dtel.html be used instead?

Comment (mbtools) [reply src/objects/sap/zif_abapgit_sap_report.intf.abap]: done

File: src/ui/pages/zcl_abapgit_gui_page_debuginfo.clas.abap [hunk 2]
```diff
@@ -20,7 +20,7 @@ CLASS zcl_abapgit_gui_page_debuginfo DEFINITION
   PROTECTED SECTION.
   PRIVATE SECTION.
 
-    CONSTANTS c_exit_standalone TYPE c LENGTH 30 VALUE 'ZABAPGIT_USER_EXIT' ##NO_TEXT.
+    CONSTANTS c_exit_standalone TYPE programm VALUE 'ZABAPGIT_USER_EXIT' ##NO_TEXT.
```
Comment (larshp) [inline src/ui/pages/zcl_abapgit_gui_page_debuginfo.clas.abap:23]:
```suggestion
    CONSTANTS c_exit_standalone TYPE syrepid VALUE 'ZABAPGIT_USER_EXIT' ##NO_TEXT.
```

File: src/objects/oo/zcl_abapgit_oo_serializer.clas.abap [hunk 3]
```diff
@@ -130,7 +132,8 @@ CLASS zcl_abapgit_oo_serializer IMPLEMENTATION.
   METHOD read_include.
 
     DATA: ls_include TYPE progstruc.
-
+    DATA lv_program TYPE programm.
```
Comment (larshp) [inline src/objects/oo/zcl_abapgit_oo_serializer.clas.abap:135]:
```suggestion
    DATA lv_program TYPE syrepid.
```

**Comment by mbtools** [comment]
I replaced all remaining `programm` with `syrepid`

**Comment by larshp** [comment]
nice, unit tests now passing

**Review by larshp — APPROVED** [review]

**Comment by chrassig** [comment]
We encountered syntax errors after pulling the current version of abapGit in a system with ABAP version 750 and transporting abapGit to a system with ABAP version 740 (a Solution Manager system, to my knowledge, there is no release of Solution Manager with an ABAP version newer than 740).

> Program ZCL_ABAPGIT_SAP_REPORT========CP, Include ZCL_ABAPGIT_SAP_REPORT========CM004: Syntax error in line 000020
> Unable to interpret 'VERSION'. Possible causes oferror: Incorrect spelling or comma error 

In ABAP release 7.40, the statement INSERT REPORT does not support the addition VERSION that is now used in class ZCL_ABAPGIT_SAP_REPORT of abapGit:

> For this reason, the addition UNICODE ENABLING of the statement INSERT REPORT has been replaced by the universal addition VERSION and is now obsolete"

https://help.sap.com/doc/abapdocu_750_index_htm/7.50/en-US/abennews-750-abap_versions.htm#!ABAP_MODIFICATION_2@2@

Is it still the goal of abapGit to support ABAP versions as old as 702? Or will we have to look for a different update and transport strategy for abapGit in our SAP system landscape in the future, as support for older releases will be dropped?

I was able to get a version of abapGit without syntax errors by switching back to tag v1.124.0.

**Comment by chrassig** [comment]
I have been looking for a few minutes for a syntax that would allow systems with newer ABAP version (750 or higher) to use the INSERT REPORT statement with the VERSION addition, while enabling systems with older ABAP versions (lower than 750) to still import abapGit via a transport request without running into syntax check.

For example, I was looking for a function module or global class in SAP coding that encapsulates the INSERT REPORT statement including the VERSION addition. We could call a function module or global class pseudo-dynamically if it exists.

Unfortunately, I did not find a matching function module or a class yet, and I do not have a different working idea yet.

**Comment by larshp** [comment]
@chrassig please open a new issue

**Comment by chrassig** [comment]
I saw that Marc (@mbtools) already changed the code. The VERSION addition is not used any more.

@larshp, do you still want me to open a new issue?

Thanks a lot to both of you!

**Comment by larshp** [comment]
no, but for future reference, its really difficult to keep track of issues inside closed issues/PRs...

**Comment by chrassig** [comment]
I agree. With my first comment, I just meant to ask what to do.

I am extremely grateful that you guys invest your time into this project, and I want to try to do whatever I can to contribute and cause you as little work as possible.

So opening an issue would not be a problem for me at all. I just had the impression that it might do more harm than good to open an issue for a problem that has already been fixed.

**Comment by mbtools** [comment]
If something isn't working, issues are the way to go since they document it for others (including your future self). 

If you or your company really feel bad about it (or love abapGit so much), you can sponsor us 😉

**Pull request merged**